### ISSUE TITLE
Inconsistent answer from read_group with duplicate group keys, leading to flux test failure

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#2681]
https://github.com/influxdata/flux/blob/master/stdlib/planner/group_first_last_test.flux is failing when run against IOX

# Reproducer:

Input looks like this
```
// two fields, two tags keys, with three rows in each combo
inData = array.from(
    rows: [
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 3},
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:40Z, _value: 1},
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:50Z, _value: 0},
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 4},
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:40Z, _value: 3},
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:50Z, _value: 1},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 1},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:40Z, _value: 0},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:50Z, _value: 4},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 4},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:40Z, _value: 0},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:50Z, _value: 4},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 0},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:40Z, _value: 0},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:50Z, _value: 0},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 0},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:40Z, _value: 4},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:50Z, _value: 3},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 3},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:40Z, _value: 2},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:50Z, _value: 1},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 1},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:40Z, _value: 0},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:50Z, _value: 2},
    ],
)
    |> group(columns: ["_measurement", "_field", "t0", "t1"])
```

 query is
```
    got = testing.load(tables: inData)
        |> range(start: -100y)
        |> group(columns: ["t0"])
        |> first()
        |> drop(columns: ["_start", "_stop"])
```

Expected output looks like this:
```
        rows: [
            {_measurement: "m0", _field: "f0", "t0": "t0v0", "t1": "t1v0", "_value": 3, _time: 2021-07-06T23:06:30Z},
            {_measurement: "m0", _field: "f0", "t0": "t0v1", "t1": "t1v0", "_value": 1, _time: 2021-07-06T23:06:30Z},
        ],
```

The diff reported by `fluxtest` is:
```
group_one_tag_first...fail: # t0=t0v0 _diff=string,_field=string,_measurement=string,_time=time,_value=int,t1=string
t0=t0v0 _diff=-,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=3i,t1=t1v0
t0=t0v0 _diff=+,_field=f1,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=0i,t1=t1v1
# t0=t0v1 _diff=string,_field=string,_measurement=string,_time=time,_value=int,t1=string
t0=t0v1 _diff=-,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=1i,t1=t1v0
t0=t0v1 _diff=+,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=4i,t1=t1v1
```

**Comment by alamb** [comment]
I think I know what is going on here -- working on a writeup

**Comment by alamb** [comment]
# Analysis of the difference: 

TLDR; I think IOx and TSM are resolving "ties" on the group key / timestamp differently, and that we should update IOx to follow the TSM behavior

Note there are several rows that have `_time=2021-07-06T23:06:30Z` and the same value of the `t0` tag. TSM returns the row that has `t0=t0v0, t1=t1v0 @ 2021-07-06T23:06:30Z` and `t0=t0v1, t1=t1v0 @ 2021-07-06T23:06:30Z` whereas IOx seems to pick the different `t1` value.


In pictures, if you look at all the rows with `_time: 2021-07-06T23:06:30Z` you can see there are *two* distinct values for t1 for each distinct value of `t0`:
```
inData = array.from(
    rows: [
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 3}, <-- TSM (t0=t0v0)
        {_field: "f0", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 4},
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 1}, <-- TSM (t0=t0v1)
        {_field: "f0", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 4}, <-- IOX (t0=t0v1)
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 0},
        {_field: "f1", _measurement: "m0", t0: "t0v0", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 0}, <-- IOX (t0=t0v0
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v0", _time: 2021-07-06T23:06:30Z, _value: 3},
        {_field: "f1", _measurement: "m0", t0: "t0v1", t1: "t1v1", _time: 2021-07-06T23:06:30Z, _value: 1},
    ],
)
```

Ala this is like the equivalent of writing SELECT FIRST(time) OVER (PARTITION by t0) and there are ties for t0.

My suspicion is that TSM happens to iterate through the values in its native tagset order (aka the data is lexographically ordered by tags), whereas the IOx implementation will pick whatever particular row happens to come out of the `HashAggregateExec`

```
  HashAggregateExec: mode=Final, gby=[t0@0 as t0, t1@1 as t1], aggr=[f0, time_f0, f1, time_f1]
    CoalescePartitionsExec
      HashAggregateExec: mode=Partial, gby=[t0@2 as t0, t1@3 as t1], aggr=[f0, time_f0, f1, time_f1]
        CoalesceBatchesExec: target_batch_size=500
          FilterExec: -1522736144178816000 <= time@4 AND time@4 < 1633023855821184000
            RepartitionExec: partitioning=RoundRobinBatch(16)
              DeduplicateExec: [t0@2 ASC,t1@3 ASC,time@4 ASC]
                SortExec: [t0@2 ASC,t1@3 ASC,time@4 ASC]
                  IOxReadFilterNode: table_name=m0, chunks=1 predicate=Predicate exprs: [TimestampNanosecond(-1522736144178816000) <= #time, #time < TimestampNanosecond(1633023855821184000)]
```


So while according to the [MiniSpec for read_group](https://docs.google.com/document/d/1UoCov4_NEhJHTxQftn5-YlzfhCiP5hgqDq9ikbRRa3g/edit#) we could quibble about the test case being bad (that its semantics aren't precisely defined) given we don't know who might be relying on this behavior, I think the safest thing to do is ensure the sort order prior to calculating aggregates is the same as TSM so that ties are resolved the same way

**Comment by alamb** [comment]
Before I code up this change, @e-dard  could you give it a once over to make sure I am not doing something crazy?

**Comment by e-dard** [comment]
I will look at this in detail first thing. 

But I think you are on the right lines. Good write up!

**Comment by e-dard** [comment]
Ok here is what I see. TLDR I agree, but I thought I would write some of this up for the benefit of others.

The input data has *eight* distinct series keys. In InfluxDB Cloud the line protocol would be manipulated to include a `_measurement` tag and a `_field` tag, but I'll just list them out without manipulating the tags around. The points for the same series are stored in the order below too (ordered by time).

```
m0,t0=t0v0,t1=t1v0 f0=3 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v0 f0=1 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v0 f0=0 2021-07-06T23:06:50Z

m0,t0=t0v0,t1=t1v0 f1=0 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v0 f1=0 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v0 f1=0 2021-07-06T23:06:50Z

m0,t0=t0v0,t1=t1v1 f0=4 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v1 f0=3 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v1 f0=1 2021-07-06T23:06:50Z

m0,t0=t0v0,t1=t1v1 f1=0 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v1 f1=4 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v1 f1=3 2021-07-06T23:06:50Z

m0,t0=t0v1,t1=t1v0 f0=1 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v0 f0=0 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v0 f0=4 2021-07-06T23:06:50Z

m0,t0=t0v1,t1=t1v0 f1=3 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v0 f1=2 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v0 f1=1 2021-07-06T23:06:50Z

m0,t0=t0v1,t1=t1v1 f0=4 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v1 f0=0 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v1 f0=4 2021-07-06T23:06:50Z

m0,t0=t0v1,t1=t1v1 f1=1 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v1 f1=0 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v1 f1=2 2021-07-06T23:06:50Z
```

The Flux query is:

```
|> range(start: -100y)
|> group(columns: ["t0"])
|> first()
|> drop(columns: ["_start", "_stop"])
```

Aside from the `range` filter, the first thing the Flux query does is re-group them. As we know, grouping in flux doesn't need an aggregate to run, it just works like an `order by`.

Grouping on the tag key `t0` gives us:

```
m0,t0=t0v0,t1=t1v0 f0=3 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v0 f0=1 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v0 f0=0 2021-07-06T23:06:50Z
m0,t0=t0v0,t1=t1v0 f1=0 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v0 f1=0 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v0 f1=0 2021-07-06T23:06:50Z
m0,t0=t0v0,t1=t1v1 f0=4 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v1 f0=3 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v1 f0=1 2021-07-06T23:06:50Z
m0,t0=t0v0,t1=t1v1 f1=0 2021-07-06T23:06:30Z
m0,t0=t0v0,t1=t1v1 f1=4 2021-07-06T23:06:40Z
m0,t0=t0v0,t1=t1v1 f1=3 2021-07-06T23:06:50Z

m0,t0=t0v1,t1=t1v0 f0=1 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v0 f0=0 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v0 f0=4 2021-07-06T23:06:50Z
m0,t0=t0v1,t1=t1v0 f1=3 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v0 f1=2 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v0 f1=1 2021-07-06T23:06:50Z
m0,t0=t0v1,t1=t1v1 f0=4 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v1 f0=0 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v1 f0=4 2021-07-06T23:06:50Z
m0,t0=t0v1,t1=t1v1 f1=1 2021-07-06T23:06:30Z
m0,t0=t0v1,t1=t1v1 f1=0 2021-07-06T23:06:40Z
m0,t0=t0v1,t1=t1v1 f1=2 2021-07-06T23:06:50Z
```

Notice all that happened between these two is the eight groups became two. No shuffling or re-ordering happened. This is where I imagine the `HashAggregate` is not enforcing the order.

So next we have the `first()`. Again, as we know, `first` in Flux isn't like `FIRST` in SQL. In flux, `first()` acts on the Flux table level (each group above is a Flux table). The `first` selector brings along its friends on the series key too, i.e., the other tags and timestamp. This makes understanding how it works simple. 

We take the two Flux tables (groupings) above and take the first row from each to give us:

```
m0,t0=t0v0,t1=t1v0 f0=3 2021-07-06T23:06:30Z

m0,t0=t0v1,t1=t1v0 f0=1 2021-07-06T23:06:30Z
```

Notice I wrote them separated. At this point they are still going to be treated as two Flux tables (groups) with a single entry in each.

> TLDR; I think IOx and TSM are resolving "ties" on the group key / timestamp differently, and that we should update IOx to follow the TSM behavior

I would say this Is a correct summary The key difference is that grouping in Flux is more like a "stable sort" in that the underlying time-series point ordering is respected.

So basically if we think about it as SQL, first we know that the TSM data model would get translated into the following IOx table:


| t0 | t1 | f0 | f1 | time |
| ------------- | ------------- |  ------------- | ------------- | ------------- |
| 't0v0' | 't1v0' | 3.0 |  0.0 | 2021-07-06T23:06:30Z|
| 't0v0' | 't1v0' | 1.0 |  1.0 | 2021-07-06T23:06:40Z|
| 't0v0' | 't1v0' | 0.0 |  0.0 | 2021-07-06T23:06:50Z|
| 't0v0' | 't1v1' | 4.0 |  0.0 | 2021-07-06T23:06:30Z|
| 't0v0' | 't1v1' | 3.0 |  4.0 | 2021-07-06T23:06:40Z|
| 't0v0' | 't1v1' | 1.0 |  3.0 | 2021-07-06T23:06:50Z|
| 't0v1' | 't1v0' | 1.0 |  3.0 | 2021-07-06T23:06:30Z|
| 't0v1' | 't1v0' | 0.0 |  2.0 | 2021-07-06T23:06:40Z|
| 't0v1' | 't1v0' | 4.0 |  1.0 | 2021-07-06T23:06:50Z|
| 't0v1' | 't1v1' | 4.0 | 1.0 | 2021-07-06T23:06:30Z|
| 't0v1' | 't1v1' | 0.0 | 0.0 | 2021-07-06T23:06:40Z|
| 't0v1' | 't1v1' | 4.0 | 2.0 | 2021-07-06T23:06:50Z|

And we know we want these rows:

![Screenshot 2021-10-01 at 11 11 17](https://user-images.githubusercontent.com/501993/135603515-90bbef33-0f71-40db-bab8-08d950a460a2.png)

In [SQL Equivalents of Common Flux Patterns](https://docs.google.com/document/d/1Y_Kvh-LIbAM2DEtoZ1vhlHjVpLJapZslI0pWZjPsgGY/edit#heading=h.euqp2imw0l0u) I wrote an example SQL query to do similar to the Flux query that's failing (a grouping and then a selector). 

In this case the Flux:

```
|> range(start: -100y)
|> group(columns: ["t0"])
|> first()
```

I think is semantically equivalent to the following SQL:

```sql
 SELECT *
 FROM (
                SELECT 
                ROW_NUMBER() OVER (
                        PARTITION BY "t0"
                        ORDER BY "t0", "t1" , "time"
                ) AS rn, -- keep data ordered by primary key
                *
        FROM "m0"
)t0
WHERE  t0.rn = 1; -- Acts like Flux's FIRST selector  
```

I would then expect the following two rows to be returned:

```
1	"t0v0"	"t1v0"	3	0	"2021-07-06T23:06:30Z"
1	"t0v1"	"t1v0"	1	3	"2021-07-06T23:06:30Z"
```


| rn | t0 | t1 | f0 | f1 | time |
| ------------- | ------------- | ------------- |  ------------- | ------------- | ------------- |
| 1| 't0v0' | 't1v0' | 3.0 |  0.0 | 2021-07-06T23:06:30Z|
| 1| 't0v1' | 't1v0' | 1.0 |  3.0 | 2021-07-06T23:06:30Z|

At that point I'm now wondering how we will correctly project out `f0` from the rows, given the projection (column) isn't specified within the query. 😕 ...

Ok, so here is what I think... We *know* that the `_field=fxyz` tag key is on the TSM series key. And we know that it's *always* the last lexicographically ordered tag key. So I think that means that in this case we can deduce which of the field columns to return by lexicographically ordering the non-tag (field colums).

In this case we would return the field and values `f0` from each of these rows. Perhaps you already had this figured out in the gRPC storage API 😄

**Comment by alamb** [comment]
Unassigning myself in the (not so) secret hope someone else is interested in picking this up, while I try and extract out some more failures from https://github.com/influxdata/conductor/issues/563

**Comment by alamb** [comment]
Interestingly, with the changes in https://github.com/influxdata/influxdb_iox/pull/2736 the diff is different

```
group_one_tag_first...fail: # t0=t0v0 _diff=string,_field=string,_measurement=string,_time=time,_value=int,t1=string
t0=t0v0 _diff=-,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=3i,t1=t1v0
t0=t0v0 _diff=+,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=3i,t1=!(nil)
# t0=t0v1 _diff=string,_field=string,_measurement=string,_time=time,_value=int,t1=string
t0=t0v1 _diff=-,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=1i,t1=t1v0
t0=t0v1 _diff=+,_field=f0,_measurement=m0,_time=2021-07-06T23:06:30Z,_value=1i,t1=!(nil)
```

Looking carefully, the issue is that IOx is only returning a `GroupFrame`  with the tag keys that were in the grouping expression, NOT all the tags as flux expects. I verified that when all tag keys are returned, then the fluxtest passes

Note the issue about ensuring Series come back ordered by tag values ordering is not fixed and in this particular example in my laptop the output out of the HashAggregateExec happens to be ordered correctly. 

So two issues:
1. Should group frame contain all the tag keys (not just the ones used for grouping?)
2. Ensure output is ordered in lexographical order

Will make two separate PRs

**Comment by alamb** [comment]
@e-dard  I was planning to work on ensuring the output is ordered in lexographical order today and have assigned this ticket to myself

**Comment by alamb** [comment]
🤔  so fascinatingly I have spent some time looking into this again and it turns out that this is one of the tests that is skipped in IDPE today (aka it also fails when run against TSM in IDPE). 

https://github.com/influxdata/idpe/blob/d501f15e9ca12cb89c87489efbee41fe8ce49657/scripts/ci/test/flux.bash#L34-L37

Thus I am closing this ticket, not because there isn't still a problem, but because the symptom tracked here doesn't seem to uncover an issue.
